File: blocksandstuff-blocks/src/main/java/org/everbuild/blocksandstuff/blocks/group/placement/PlacementGroup.java
package org.everbuild.blocksandstuff.blocks.group.placement;

import net.minestom.server.instance.block.Block;
import net.minestom.server.instance.block.rule.BlockPlacementRule;
import org.everbuild.blocksandstuff.blocks.group.block.BlockGroup;
import org.everbuild.blocksandstuff.blocks.group.block.IntoBlockGroup;
import org.jetbrains.annotations.NotNull;

import java.util.function.Function;

public class PlacementGroup implements IntoBlockGroup, IntoPlacementRule {
    private final BlockGroup blockGroup;
    private final Function<Block, BlockPlacementRule> valueFunction;

    public PlacementGroup(BlockGroup blockGroup, Function<Block, BlockPlacementRule> valueFunction) {
        this.blockGroup = blockGroup;
        this.valueFunction = valueFunction;
    }

    @Override
    public @NotNull BlockGroup getBlockGroup() {
        return blockGroup;
    }

    @Override
    public BlockPlacementRule createRule(Block block) {
        return valueFunction.apply(block);
    }
}

File: blocksandstuff-blocks/src/main/kotlin/org/everbuild/blocksandstuff/blocks/group/placement/PlacementGroup.kt
package org.everbuild.blocksandstuff.blocks.group.placement

import java.util.function.Function
import net.minestom.server.instance.block.Block
import net.minestom.server.instance.block.rule.BlockPlacementRule
import org.everbuild.blocksandstuff.blocks.group.block.BlockGroup
import org.everbuild.blocksandstuff.blocks.group.block.IntoBlockGroup

class PlacementGroup(
    override val blockGroup: BlockGroup,
    private val valueFunction: Function<Block, BlockPlacementRule>
) : IntoBlockGroup, IntoPlacementRule {
    override fun createRule(block: Block): BlockPlacementRule {
        return valueFunction.apply(block)
    }
}
